File: security_lab/log_analyzer.py
import re
from collections import defaultdict

# Configuration
LOG_FILE = "server.log"
FAILED_LOGIN_THRESHOLD = 3

# Regex patterns
# Handles both simple mock format and Morgan/Apache combined format
# Format 1: [2023-10-27 10:00:00] 192.168.1.1 - GET /home 200
# Format 2: 127.0.0.1 - - [23/Apr/2026:12:34:56 +0000] "GET /api/books HTTP/1.1" 200 123
LOG_PATTERN = r'(?P<ip>[\d\.a-f:A-F]+).*?\[(?P<timestamp>.*?)\].*?"?(?P<method>GET|POST|PUT|DELETE|PATCH)\s+(?P<path>\S+)(?:\s+HTTP/.*?)?"?\s+(?P<status>\d{3})'

# SQLi Detection Regex
SQLI_PATTERN = r"('|\"|%27|%22)(?:[\s\+]|%20)*(OR|UNION|SELECT|DROP|UPDATE|DELETE|--|#|/*)"

def parse_logs(file_path):
    parsed_data = []
    try:
        with open(file_path, 'r') as f:
            for line in f:
                match = re.search(LOG_PATTERN, line)
                if match:
                    parsed_data.append(match.groupdict())
    except FileNotFoundError:
        print(f"Error: {file_path} not found. Please run the log generator first.")
        return []
    return parsed_data

def detect_threats(logs):
    failed_logins = defaultdict(list)
    sqli_attempts = []
    
    for entry in logs:
        ip = entry['ip']
        path = entry['path']
        status = entry['status']
        
        # 1. Detect Failed Login (Brute Force)
        # Check for 401 or 404 on common login paths
        if (path in ["/login", "/api/auth/login", "/invalid-login"]) and (status in ["401", "404"]):
            failed_logins[ip].append(entry)
            
        # 2. Detect SQL Injection
        if re.search(SQLI_PATTERN, path, re.IGNORECASE):
            sqli_attempts.append(entry)
            
    return failed_logins, sqli_attempts

def print_security_summary(failed_logins, sqli_attempts):
    print("=" * 60)
    print("         SECURITY THREAT DETECTION REPORT")
    print("=" * 60)
    
    # Brute Force Alerts
    print(f"\n[!] BRUTE FORCE ATTACK ALERTS (Threshold: {FAILED_LOGIN_THRESHOLD} failures)")
    found_brute = False
    for ip, attempts in failed_logins.items():
        if len(attempts) >= FAILED_LOGIN_THRESHOLD:
            found_brute = True
            print(f"[-] IP: {ip} | Attempts: {len(attempts)}")
            for att in attempts[:3]: # Show first 3 for brevity
                print(f"    - [{att['timestamp']}] {att['method']} {att['path']} {att['status']}")
            if len(attempts) > 3:
                print(f"    - ... ({len(attempts) - 3} more entries)")
                
    if not found_brute:
        print("    No brute force attempts detected.")
        
    # SQLi Alerts
    print(f"\n[!] SQL INJECTION ATTEMPT ALERTS")
    if sqli_attempts:
        # Group by IP for cleaner output
        sqli_by_ip = defaultdict(list)
        for attempt in sqli_attempts:
            sqli_by_ip[attempt['ip']].append(attempt)
            
        for ip, attempts in sqli_by_ip.items():
            print(f"[-] IP: {ip} | Threats Detected: {len(attempts)}")
            for att in attempts[:3]:
                print(f"    - [{att['timestamp']}] Path: {att['path']}")
            if len(attempts) > 3:
                print(f"    - ... ({len(attempts) - 3} more entries)")
    else:
        print("    No SQL injection attempts detected.")
        
    print("\n" + "=" * 60)

def main():
    print(f"Analyzing logs from {LOG_FILE}...")
    logs = parse_logs(LOG_FILE)
    
    if not logs:
        return
        
    failed_logins, sqli_attempts = detect_threats(logs)
    print_security_summary(failed_logins, sqli_attempts)

if __name__ == "__main__":
    main()

File: security_lab/log_analyzer.js
const fs = require('fs');

const LOG_FILE = 'server.log';
const FAILED_LOGIN_THRESHOLD = 3;

// Regex: Supports simple and Morgan/Apache formats (including IPv6)
const LOG_PATTERN = /([\d\.a-f:A-F]+).*?\[(.*?)\].*?"?(GET|POST|PUT|DELETE|PATCH)\s+(\S+)(?:\s+HTTP\/.*?)?"?\s+(\d{3})/;
const SQLI_PATTERN = /('|"|%27|%22)(?:[\s\+]|%20)*(OR|UNION|SELECT|DROP|UPDATE|DELETE|--|#|\/\*)/i;

function analyze() {
  if (!fs.existsSync(LOG_FILE)) {
    console.log(`Error: ${LOG_FILE} not found. Run the app or generator first.`);
    return;
  }

  const lines = fs.readFileSync(LOG_FILE, 'utf8').split('\n');
  const failedLogins = {};
  const sqliAttempts = [];

  lines.forEach(line => {
    const match = line.match(LOG_PATTERN);
    if (match) {
      const [_, ip, timestamp, method, path, status] = match;
      const entry = { ip, timestamp, method, path, status };

      // 1. Detect Failed Login
      if (['/login', '/api/auth/login', '/invalid-login'].includes(path) && ['401', '404'].includes(status)) {
        if (!failedLogins[ip]) failedLogins[ip] = [];
        failedLogins[ip].push(entry);
      }

      // 2. Detect SQLi
      if (SQLI_PATTERN.test(path)) {
        sqliAttempts.push(entry);
      }
    }
  });

  printSummary(failedLogins, sqliAttempts);
}

function printSummary(failedLogins, sqliAttempts) {
  console.log('='.repeat(60));
  console.log('         SECURITY THREAT DETECTION REPORT (JS VERSION)');
  console.log('='.repeat(60));

  console.log(`\n[!] BRUTE FORCE ATTACK ALERTS (Threshold: ${FAILED_LOGIN_THRESHOLD} failures)`);
  let foundBrute = false;
  for (const ip in failedLogins) {
    if (failedLogins[ip].length >= FAILED_LOGIN_THRESHOLD) {
      foundBrute = true;
      console.log(`[-] IP: ${ip} | Attempts: ${failedLogins[ip].length}`);
      failedLogins[ip].slice(0, 3).forEach(att => {
        console.log(`    - [${att.timestamp}] ${att.method} ${att.path} ${att.status}`);
      });
      if (failedLogins[ip].length > 3) console.log(`    - ... (${failedLogins[ip].length - 3} more entries)`);
    }
  }
  if (!foundBrute) console.log('    No brute force attempts detected.');

  console.log(`\n[!] SQL INJECTION ATTEMPT ALERTS`);
  if (sqliAttempts.length > 0) {
    const byIp = {};
    sqliAttempts.forEach(att => {
      if (!byIp[att.ip]) byIp[att.ip] = [];
      byIp[att.ip].push(att);
    });

    for (const ip in byIp) {
      console.log(`[-] IP: ${ip} | Threats Detected: ${byIp[ip].length}`);
      byIp[ip].slice(0, 3).forEach(att => {
        console.log(`    - [${att.timestamp}] Path: ${att.path}`);
      });
      if (byIp[ip].length > 3) console.log(`    - ... (${byIp[ip].length - 3} more entries)`);
    }
  } else {
    console.log('    No SQL injection attempts detected.');
  }
  console.log('\n' + '='.repeat(60));
}

analyze();
